flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+
+        # Set the `_is_aggregate` attribute to True by default. But if the "non-aggregate"
+        # aggregator object i.e. aggregator.groupsof(1) is created using the class method,
+        # then we explicitly set the `_is_aggregate` attribute to False.
+        self._is_aggregate = True
+        self._aggregator_function = aggregator_function
+        self._sort_by = sort_by
+        self._reverse_order = bool(reverse_order)
+        self._select = select
+        # The idea of providing users with an additional tag attribute comes from the fact
+        # that the byte code of an aggregator function cannot truly differentiate it with
+        # some other aggregator function. Hence while checking the equality, we also
+        # check the equality based on tag provided by an user or us in a classmethod.
+        self._tag = tag
+
+    @classmethod
+    def groupsof(cls, num=1, sort_by=None, reverse_order=False, select=None):
+        """Aggregates jobs of a set group size.
+
+        By default aggregate of a single job is created.
+
+        If the number of jobs present in the project can't be divided by the number
+        provided by the user then the last aggregate will have the maximum possible
+        jobs in it.
+        For instace, if 10 jobs are present in a project and they are aggregated in groups
+        of 3 then the length of aggregates would be 3, 3, 3, 1.
+
+        The below code-block provides an example on how jobs can be aggregated in
+        groups of 2.
+
+        .. code-block:: python
+
+            @aggregator.groupsof(num=2)
+            @FlowProject.operation
+            def foo(*jobs):
+                print(len(jobs))
+
+        :param num:
+            The default size of aggregates excluding the final aggregate.
+        :type num:
+            int
+        :param sort_by:
+            Before aggregating, sort the jobs by a given statepoint parameter.
+            The default behavior is no sorting.
+        :type sort_by:
+            str or NoneType
+        :param reverse_order:
+            States if the jobs are to be sorted in reverse order.
+            The default value is False.
+        :type reverse_order:
+            bool
+        :param select:
+            Condition for filtering individual jobs. This is passed as the
+            callable argument to `filter`.
+            The default behavior is no filtering.
+        :type select:
+            callable or NoneType
+        """
+        try:
+            if num != int(num):
+                raise ValueError('The num parameter should be an integer')
+            num = int(num)
+            if num <= 0:
+                raise ValueError('The num parameter should have a value greater than 0')
+        except TypeError:
+            raise TypeError('The num parameter should be an integer')
+
+        def aggregator_function(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
```

Comment:
This solution is pretty hacky... it's abusing the fact that you can store multiple references to the same iterator and then iterate over each reference in sequence to produce chunks of the original. It certainly works, though, and maybe it is the most efficient way to do this without making copies. I assume you found this recipe somewhere, could you add a link to document that, and maybe a comment here as well describing how it works?

Reply:
This was originally implemented by @csadorf in #52 and that had a link from where it was copied.
I'll add the link into this PR and also mention #52